Zero-initialize _data in Image::resize(), don't leave _data dangling in Image::operator=()

oleg-derevenetz opened the pull request:
fix #3145

std::vector::resize() performs value-initialization (zero-initialization for primitive types) of new members, while new[] doesn't. Also I wonder, why not use std::unique_ptr<T[]> for _data instead of raw pointer? There is no overhead in terms of memory usage, but there will be no room for dangling references.

ihhub commented on src/engine/image.cpp:
```diff
@@ -441,6 +441,9 @@ namespace fheroes2
                 _data = new uint8_t[static_cast<size_t>( _width * _height * 2 )];
             }
+            else {
+                _data = nullptr;
```
This is a good catch! I forgot to assign it to nullptr.

ihhub commented on src/engine/image.cpp:
```diff
@@ -506,6 +509,7 @@ namespace fheroes2
 
         const size_t totalSize = static_cast<size_t>( _width * _height );
         _data = new uint8_t[totalSize * 2];
+        std::fill( _data, _data + totalSize * 2, 0 );
```
We mustn't do this at all. It affects performance. The previous change was exactly to address performance issues.

oleg-derevenetz replied:
Hi @ihhub I understand the idea. The problem here is that may not even know we're working with uninitialized data: everything may be fine by coincidence - we **almost** always get zero-initialized memory... until one day we don't, and get a bug that is difficult to reproduce. It may be something to think about, so that when trying to work with uninitialized data we get an assert or an exception? May be some additional flag or keep _data nullptr until we know that it's properly initialized (and check for this)?

ihhub replied:
The problem is that we return a pointer to data in image functions and it's almost impossible to track the usage.

oleg-derevenetz replied:
@ihhub Yes, I understand - the heavy legacy of existing code :) Well, let's close this PR without merging, there is room for thought...

ihhub replied:
Okay, we could do some assert in const methods of `image` and `transform`. It might help to detect all these memory related cases but then we have other non-const methods and a lot of code... Not a trivial thing to do :(

Could you please address then other place which I pointed for now and probably add std::unique_ptr usage. Let me think how we can deal with potential issues with unitialized images

ihhub replied:
The fun fact is that the engine was actually written with an idea in mind not to zero data at all and I slipped the fact that std::vector::resize does zeroying :(

oleg-derevenetz replied:
@ihhub Yes, of course. I propose to do a to PRs though - one with nullptr and backsf.reset(), and one with unique_ptr to not mix different changes and to make it easier to rollback the changes.

ihhub replied:
Sure! Let's do it.

ihhub commented:
The original issue should be fixed by adding `backsf.reset();` at line 727 in **buildinginfo.cpp** file (**DwellingsBar** constructor).

The change made recently was to improve overall rendering speed as everything is rendered by CPU so we would like to avoid extra operations by CPU. Also since most of images we just load from AGG file there was no need to even initialize images to zeroes as according to `decodeICNSprite` we actually call **sprite.reset();** which resets the transform layer to skip pixels and image layer to zeroes. So we were doing double array reset.

If you think that std::unique_ptr<T[]> will help with code readability and won't introduce memory leaks (I remember that some implementations of std::unique_ptr don't work properly with arrays) then it's totally fine to change it.

ihhub commented:
Another explanation why we mustn't do zeroing in resizing: we are creating a new image and immediately copying the result there.

I hope that my explanation is satisfying :)

ihhub commented:
@oleg-derevenetz , actually, let's keep your std::fill call here for zeroing. I think at least we will have an explicit call which we surely won't forget and close to 1.0 version we might something better to catch these errors. My apologies for changing mind so quickly.

ihhub commented:
@oleg-derevenetz , good job! In the future we should remove this zeroing to make the application run faster.

oleg-derevenetz commented:
@ihhub I added a TODO comment regarding this and added call of `backsf.reset()` in DwellingsBar initialization as well. I'll try to implement `std::unique_ptr` as a separate PR later today.

ihhub commented:
@oleg-derevenetz , excellent! I already found another place of unitialized data (without zeroing)... -_-

oleg-derevenetz commented:
@ihhub Well, I have no doubt that there will be a couple of such places :)

ihhub commented:
Haha, let's see. I'll test all of this and ask @Branikolog to help as well with a special version having an assertion.
